services/src/api/model/db_types.rs

```diff
@@ -607,6 +617,928 @@ impl<S: std::hash::BuildHasher + std::default::Default> From<HashMapTextTextDbTy
+            } => Ok(Self::StartDuration {
+                start_field: start_duration.start_field,
+                start_format: start_duration.start_format,
+                duration_field: start_duration.duration_field,
+            }),
+            _ => Err(Error::UnexpectedInvalidDbTypeConversion),
+        }
+    }
+}
+
+#[derive(Debug, ToSql, FromSql)]
+#[postgres(name = "FormatSpecifics")]
+pub struct FormatSpecificsDbType {
+    csv: Option<FormatSpecificsCsvDbType>,
+}
+
+#[derive(Debug, ToSql, FromSql)]
+#[postgres(name = "FormatSpecificsCsv")]
+pub struct FormatSpecificsCsvDbType {
+    header: CsvHeader,
+}
+
+impl From<&FormatSpecifics> for FormatSpecificsDbType {
+    fn from(other: &FormatSpecifics) -> Self {
+        match other {
+            FormatSpecifics::Csv { header } => Self {
+                csv: Some(FormatSpecificsCsvDbType { header: *header }),
+            },
+        }
+    }
+}
+
+impl TryFrom<FormatSpecificsDbType> for FormatSpecifics {
+    type Error = Error;
+
+    fn try_from(other: FormatSpecificsDbType) -> Result<Self, Self::Error> {
+        match other {
+            FormatSpecificsDbType {
+                csv: Some(FormatSpecificsCsvDbType { header }),
+            } => Ok(Self::Csv { header }),
+            _ => Err(Error::UnexpectedInvalidDbTypeConversion),
+        }
+    }
+}
+
+#[derive(Debug, ToSql, FromSql)]
+#[postgres(name = "OgrSourceColumnSpec")]
+pub struct OgrSourceColumnSpecDbType {
+    pub format_specifics: Option<FormatSpecifics>,
+    pub x: String,
+    pub y: Option<String>,
+    pub int: Vec<String>,
+    pub float: Vec<String>,
+    pub text: Vec<String>,
+    pub bool: Vec<String>,
+    pub datetime: Vec<String>,
+    pub rename: Option<HashMapTextTextDbType>,
+}
+
+impl From<&OgrSourceColumnSpec> for OgrSourceColumnSpecDbType {
+    fn from(other: &OgrSourceColumnSpec) -> Self {
+        Self {
+            format_specifics: other.format_specifics.clone(),
+            x: other.x.clone(),
+            y: other.y.clone(),
+            int: other.int.clone(),
+            float: other.float.clone(),
+            text: other.text.clone(),
+            bool: other.bool.clone(),
+            datetime: other.datetime.clone(),
+            rename: other.rename.as_ref().map(Into::into),
+        }
+    }
+}
+
+impl TryFrom<OgrSourceColumnSpecDbType> for OgrSourceColumnSpec {
+    type Error = Error;
+
+    fn try_from(other: OgrSourceColumnSpecDbType) -> Result<Self, Self::Error> {
+        Ok(Self {
+            format_specifics: other.format_specifics,
+            x: other.x,
+            y: other.y,
+            int: other.int,
+            float: other.float,
+            text: other.text,
+            bool: other.bool,
+            datetime: other.datetime,
+            rename: other.rename.map(Into::into),
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct PolygonRef<'p> {
+    pub rings: &'p [Vec<Coordinate2D>],
+}
+
+#[derive(Debug)]
+pub struct PolygonOwned {
```

Comment:
isn't this the same as a MultiLineString? Can you reuse it?

Reply:
No, for `MultiLineString` I use and open path, but polygonal paths are closed.